### ISSUE TITLE
Small suggested clarifications

### ISSUE THREAD

**Issue by rsnikhil** [riscv-non-isa/riscv-trace-spec#138]
The word "E-Trace", used on pages 9, 10 and 79, presumably stands for "Efficient Trace", but AFAIK is nowhere so defined.

Suggest adding an item in Section 1.1 Terminology for this.

**Comment by rsnikhil** [comment]
‘ucause’ is mentioned in Section 3.2.3; Section 4.1 second bullet; table 4.
‘utval’ is mentioned in Section 4.1 second bullet; Table 4; Table 17; 7.3.2.

AFAIK these CSRs were removed from the Privileged ISA spec?

**Comment by rsnikhil** [comment]
In Section 4.1, we have:

    The following information is mandatory:
    ...
    * The instruction_type of retired instructions for:
      * Jumps with a target that cannot be inferred from the source code;
    ...

and

    The following information is optional:
    ...
    * The instruction_type of instructions for:
      ...
      * Jumps with a target that cannot be inferred from the source code;

This seems contradictory (both mandatory and optional)?

**Comment by IainCRobertson** [comment]
Rishiyur,
                That's a good observation, and does indeed appear contradictory.  It came about as a result of an earlier clarification attempt.  The list of optional types in section 4.1 corresponds to the more detailed classifications in section 4.11.  Originally, this was broken down into Calls, Tail-calls, Returns, Co-routine swap and Other.  However, there is no formal definition of tail-call in the RISC-V ISA, and so references to 'tail call' in both section 4.1.1 and 4.1 were replaced with 'jump'.  With hindsight, this is too general, and probably needs to be something like "jump with discarded link address" or "call with discarded link address"

Please open an issue against the spec (https://github.com/riscv-non-isa/riscv-trace-spec/issues) and I'll address this in the next update.

Iain

**Comment by IainCRobertson** [comment]
FWIW, when CTR went through ARC review, ARC requested the following
definitions for the various control transfer types:

[image: image.png]

**Comment by IainCRobertson** [comment]
Beeman,
                Thanks – that looks like the terminology we should adopt.  Unfortunately your screen-grab doesn’t show up in the GIT issue – can you paste plain text or include a reference that I can follow later please?

Iain

**Comment by IainCRobertson** [comment]
Sure, I'll paste the adoc text.

**Comment by IainCRobertson** [comment]
It was also improved in N-Trace.  See this:

https://github.com/riscv-non-isa/tg-nexus-trace/blob/main/docs/RISC-V-N-Trace.adoc#2-trace-ingress-port

IMPORTANT: I am against that PR mentioned below.

We agreed to EXTRACT this into SINGLE source (as fast-track).

Thanks,
Robert


________________________________
From: ***@***.*** ***@***.***> on behalf of Mr Iain Robertson via lists.riscv.org ***@***.***>
Sent: Wednesday, September 4, 2024 8:40 AM
To: Beeman Strong ***@***.***>; ***@***.*** ***@***.***>
Cc: riscv-non-isa/riscv-trace-spec ***@***.***>; riscv-non-isa/riscv-trace-spec ***@***.***>; Subscribed ***@***.***>
Subject: [EXTERNAL]Re: [RISC-V] [sig-debug-trace-perf-mon] [riscv-non-isa/riscv-trace-spec] Small suggested clarifications (Issue #138)


Beeman,

                Thanks – that looks like the terminology we should adopt.  Unfortunately your screen-grab doesn’t show up in the GIT issue – can you paste plain text or include a reference that I can follow later please?



Iain

**Comment by IainCRobertson** [comment]
I see the issue hasn't been filed yet, so pasting the text here:

|INDCALLINH |Inhibit recording of indirect calls.  See <<Transfer Type
Filtering>>.

|DIRCALLINH |Inhibit recording of direct calls.  See <<Transfer Type
Filtering>>.

|INDJMPINH |Inhibit recording of indirect jumps (without linkage).  See
<<Transfer Type Filtering>>.

|DIRJMPINH |Inhibit recording of direct jumps (without linkage).  See
<<Transfer Type Filtering>>.

|CORSWAPINH |Inhibit recording of co-routine swaps.  See <<Transfer Type
Filtering>>.

|RETINH |Inhibit recording of function returns.  See <<Transfer Type
Filtering>>.

|INDLJMPINH |Inhibit recording of other indirect jumps (with linkage).  See
<<Transfer Type Filtering>>.

|DIRLJMPINH |Inhibit recording of other direct jumps (with linkage).  See
<<Transfer Type Filtering>>.

**Comment by IainCRobertson** [comment]
Robert,
             The issue is for a request to clarify chapter 4.  Whether that is done as part of an incremental update to the E-Trace spec (along with some other clarifications and corrections) before we pull out chapter 4, or whether chapter 4 gets pulled out first and the clarifications are handled as part of that makes no difference – the important thing is that we have a record of the need for clarification.

Iain

**Comment by IainCRobertson** [comment]
N-Trace has 'correct' (ARC-blessed ...) names.
We (me, Jay, Ved=ARC) spent a lot of time (several meetings with ARC) making 'itype' names correctly named and correctly explained.

I assume 'extracted spec' will inherit this.

Thanks,
Robert
________________________________
From: Robertson, Iain ***@***.***>
Sent: Wednesday, September 4, 2024 9:50 AM
To: Robert Chyla ***@***.***>; Beeman Strong ***@***.***>; ***@***.*** ***@***.***>
Cc: riscv-non-isa/riscv-trace-spec ***@***.***>; riscv-non-isa/riscv-trace-spec ***@***.***>; Subscribed ***@***.***>
Subject: RE: [EXTERNAL]Re: [RISC-V] [sig-debug-trace-perf-mon] [riscv-non-isa/riscv-trace-spec] Small suggested clarifications (Issue #138)


Robert,

             The issue is for a request to clarify chapter 4.  Whether that is done as part of an incremental update to the E-Trace spec (along with some other clarifications and corrections) before we pull out chapter 4, or whether chapter 4 gets pulled out first and the clarifications are handled as part of that makes no difference – the important thing is that we have a record of the need for clarification.



Iain

**Comment by IainCRobertson** [comment]
Robert,
              Yes, and those are the same names Beeman put forward.  Again, no reason not to update the E-Trace spec if it happens that an update is done before chapter 4 is pulled out.

Iain

**Comment by IainCRobertson** [comment]
@rsnikhil please review https://github.com/riscv-non-isa/riscv-trace-spec/pull/161 - does this address all of your concerns?

**Comment by rsnikhil** [comment]
Yes, all these changes look fine to me (my apologies for my delay in reviewing).
